internal/storage/storageutil/client.go

```diff
@@ -70,11 +72,69 @@ type StorageClientConfig struct {
 	TracingEnabled bool
 }
 
+var (
+	// dnsCache stores a single, atomically-updated dnsCacheEntry.
+	dnsCache      atomic.Value
+	dnsCacheTTL   = 10 * time.Minute // Time-to-live for DNS cache entries.
+	defaultDialer = &net.Dialer{}    // The underlying dialer used for network connections.
+)
+
+type dnsCacheEntry struct {
+	host       string
+	addrs      []string
+	expiration time.Time
+}
+
+func dialContextWithDNSCache(ctx context.Context, network, addr string) (net.Conn, error) {
+	host, port, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to split host and port for address %q: %w", addr, err)
+	}
+
+	// Atomically load the cached entry.
+	cached := dnsCache.Load()
+	var entry dnsCacheEntry
+
+	// Check if the cache is valid for the current host and not expired.
+	if cached != nil {
+		entry = cached.(dnsCacheEntry)
+		if entry.host == host && time.Now().Before(entry.expiration) {
```

Comment:
I know in GCS probably it's always the same host. but coud you add a log if the cached entry is not the same as host? So we know something unexpected is happening.